Framework/DataHandling/src/H5Util.cpp

```diff
@@ -0,0 +1,306 @@
+  return H5Util::getDataSpace(data.size());
+}
+
+DSetCreatPropList setCompressionAttributes(const std::size_t length,
+                                           const int deflateLevel) {
+  DSetCreatPropList propList;
+  hsize_t chunk_dims[1] = {length};
+  propList.setChunk(1, chunk_dims);
+  propList.setDeflate(deflateLevel);
+  return propList;
+}
+
+void writeStrAttribute(Group &location, const std::string &name,
+                       const std::string &value) {
+  StrType attrType(0, H5T_VARIABLE);
+  DataSpace attrSpace(H5S_SCALAR);
+  auto groupAttr = location.createAttribute(name, attrType, attrSpace);
+  groupAttr.write(attrType, value);
+}
+
+void write(Group &group, const std::string &name, const std::string &value) {
+  StrType dataType(0, value.length() + 1);
+  DataSpace dataSpace = getDataSpace(1);
+  H5::DataSet data = group.createDataSet(name, dataType, dataSpace);
+  data.write(value, dataType);
+}
+
+template <typename NumT>
+void writeArray1D(Group &group, const std::string &name,
+                  const std::vector<NumT> &values) {
+  DataType dataType(getType<NumT>());
+  DataSpace dataSpace = getDataSpace(values);
+
+  DSetCreatPropList propList = setCompressionAttributes(values.size());
+
+  auto data = group.createDataSet(name, dataType, dataSpace, propList);
+  data.write(&(values[0]), dataType);
+}
+
+// -------------------------------------------------------------------
+// read methods
+// -------------------------------------------------------------------
+
+std::string readString(H5::H5File &file, const std::string &path) {
+  try {
+    auto data = file.openDataSet(path);
+    return readString(data);
+  } catch (H5::FileIException &e) {
+    UNUSED_ARG(e);
+    return "";
+  } catch (H5::GroupIException &e) {
+    UNUSED_ARG(e);
+    return "";
+  }
+}
+
+std::string readString(H5::Group &group, const std::string &name) {
+  try {
+    auto data = group.openDataSet(name);
+    return readString(data);
+  } catch (H5::GroupIException &e) {
+    UNUSED_ARG(e);
+    return "";
+  }
+}
+
+std::string readString(H5::DataSet &dataset) {
+  std::string value;
+  dataset.read(value, dataset.getDataType(), dataset.getSpace());
+  return value;
+}
+
+template <typename NumT>
+std::vector<NumT> readArray1DCoerce(H5::Group &group, const std::string &name) {
+  std::vector<NumT> result;
+
+  try {
+    DataSet dataset = group.openDataSet(name);
+    result = readArray1DCoerce<NumT>(dataset);
+  } catch (H5::GroupIException &e) {
+    UNUSED_ARG(e);
+    g_log.information() << "Failed to open dataset \"" << name << "\"\n";
+  } catch (H5::DataTypeIException &e) {
+    UNUSED_ARG(e);
+    g_log.information() << "DataSet \"" << name << "\" should be double"
+                        << "\n";
+  }
+
+  return result;
+}
+
+namespace {
+// correct for widening - narrowing should use something else
+template <typename InputNumT, typename OutputNumT>
```

Comment:
Since `wideningRead` and `narrowingRead` only differ in the lambda expression, would you be better off writing one function and passing the lambda as a parameter?